avm/src/daemon/stats.rs

```diff
@@ -0,0 +1,176 @@
+use futures::future::join_all;
+use futures::stream::StreamExt;
+use heim_common::units::{information::kilobyte, time::second};
+#[cfg(target_os = "linux")]
+use heim_cpu::os::linux::CpuTimeExt;
+#[cfg(target_os = "linux")]
+use heim_memory::os::linux::MemoryExt;
+use serde::Serialize;
+use tokio::time::{sleep, Duration};
+
+#[derive(Debug, Serialize)]
+pub struct CPUSecsV1 {
+  user: f64,
+  system: f64,
+  idle: f64,
+  irq: f64,
+  nice: f64,
+  ioWait: f64,
+  softIrq: f64,
+  steal: f64,
+}
+
+#[allow(non_snake_case)]
+#[derive(Debug, Serialize)]
+pub struct VMStatsV1 {
+  cpuSecs: Vec<CPUSecsV1>,
+  procsCpuUsage: Vec<f64>,
+  totalMemoryKb: u64,
+  availableMemoryKb: u64,
+  freeMemoryKb: u64,
+  usedMemoryKb: u64,
+  activeMemoryKb: u64,
+  totalSwapKb: u64,
+  usedSwapKb: u64,
+  freeSwapKb: u64,
+}
+
+// calculate the cpu times usage per process using the process'
+// total cpu time delta in a given time window
+async fn get_proc_usages() -> Vec<f64> {
+  let duration = 1.0;
+  let futures = heim_process::processes()
+    .map(|process| async {
+      match process {
+        Ok(proc) => {
+          let cpu_1 = proc.cpu_time().await;
+          sleep(Duration::from_secs(duration as u64)).await;
+          let cpu_2 = proc.cpu_time().await;
+          // account for zombie process
+          if cpu_1.is_err() || cpu_2.is_err() {
+            return 0.0;
+          }
+          let times_1 = cpu_1.unwrap();
+          let times_2 = cpu_2.unwrap();
+          let system = times_2.system().get::<second>() - times_1.system().get::<second>();
+          let user = times_2.user().get::<second>() - times_1.user().get::<second>();
+          (user + system) / duration
+        }
+        Err(_) => 0.0,
```

Comment:
this cpu usage per process calculation does the right thing and replaces the one in https://github.com/alantech/alan/pull/457/files#r597315536

Reply:
It definitely looks good, but why divide by the duration if it is 1.0?

Reply:
oh true. I did that just in case we changed this value to a duration other than one second